psydac/utilities/utils.py

```diff
@@ -5,13 +5,16 @@
 import numpy as np
 
 #===============================================================================
-def refine_array_1d( x, n ):
+def refine_array_1d(x, n, remove_duplicates=True):
 
     xr = []
-    for (a,b) in zip(x[:-1],x[1:]):
-        xr.extend( np.linspace( a, b, n, endpoint=False ) )
-    xr.append( x[-1] )
-
+    if not remove_duplicates:
+        if n == 1:
+            raise ValueError("Cannot have duplicate points with n = 1")
+    for (a, b) in zip(x[:-1], x[1:]):
+        xr.extend(np.linspace(a, b, n, endpoint=not remove_duplicates))
+    if remove_duplicates:
+        xr.append(x[-1])
     return xr
```

Comment:
Unfortunately there is no docstring, so it was not clear what the argument `n` represents.

When I wrote this function I meant `n` to be the number of subdivisions within each cell, and I think that assumption should be kept. So in the case where `remove_duplicates == False` we should pass `n+1` to `linspace`. Further, `n==1` is an acceptable value in all cases.

Reply:
If we do that then the `refinement` parameters and `n` are different which is why I did that.

Reply:
There is nothing wrong if the two parameters are different. But to avoid confusion we may want to use a different name instead of `refinement`. Something like `npts_per_cell` is more explicit.

Reply:
Fixed in [3d32ea6](https://github.com/pyccel/psydac/pull/193/commits/3d32ea6ea706b76e80e6d0f74ca8acf36c47be3b)

Reply:
Thanks!

Could you please add a docstring to this function? I think you know how it works now 😉

Reply:
I already did in 3d32ea6 or I am not understanding which function  you are referring to.

Reply:
Sorry, I was looking at the conversation and the code was outdated... I did not realize that you added the docstring! I am going to suggest some improvements, though